substrate/frame/assets-freezer/src/lib.rs

```diff
@@ -0,0 +1,189 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! # Assets Freezer Pallet
+//!
+//! A pallet capable of freezing fungibles from `pallet-assets`.
+//!
+//! > Made with *Substrate*, for *Polkadot*.
+//!
+//! [![github]](https://github.com/paritytech/polkadot-sdk/tree/master/substrate/frame/examples/basic)
+//! [![polkadot]](https://polkadot.network)
+//!
+//! [polkadot]: https://img.shields.io/badge/polkadot-E6007A?style=for-the-badge&logo=polkadot&logoColor=white
+//! [github]: https://img.shields.io/badge/github-8da0cb?style=for-the-badge&labelColor=555555&logo=github
+//!
+//! ## Pallet API
+//!
+//! See the [`pallet`] module for more information about the interfaces this pallet exposes,
+//! including its configuration trait, dispatchables, storage items, events and errors.
+//!
+//! ## Overview
+//!
+//! This pallet provides the following functionality:
+//!
+//! - Pallet hooks allowing [`pallet-assets`] to know the frozen balance for an account on a given
+//!   asset (see [`pallet_assets::FrozenBalance`]).
+//! - An implementation of
+//!   [`fungibles::InspectFreeze`)(frame_support::traits::fungibles::InspectFreeze) and
+//!   [`fungibles::MutateFreeze`](frame_support::traits::fungibles::MutateFreeze), allowing other
+//!   pallets to manage freezes for the `pallet-assets` assets.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{tokens::IdAmount, VariantCount},
+	BoundedVec,
+};
+use frame_system::pallet_prelude::BlockNumberFor;
+use pallet_assets::{AccountIdOf, AssetBalanceOf, AssetIdOf};
+use sp_runtime::{
+	traits::{Saturating, Zero},
+	BoundedSlice,
+};
+
+pub use pallet::*;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+mod impls;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::config(with_default)]
+	pub trait Config<I: 'static = ()>: frame_system::Config + pallet_assets::Config<I> {
+		/// The ID type for freezes.
+		type FreezeIdentifier: Parameter + Member + MaxEncodedLen + Copy;
+
+		/// The overarching freeze reason.
+		#[pallet::no_default_bounds]
+		type RuntimeFreezeReason: VariantCount;
```

Comment:
can you tell please why two separate types?

Reply:
This follows the logic present in `pallet-balances`: 

https://github.com/paritytech/polkadot-sdk/blob/0d71753e0eb1c0fdbeb09af5db7a2e1b5f150b13/substrate/frame/balances/src/lib.rs#L250

This is used to constraint the maximum amount of freezes to the count of variants in `RuntimeFreezeReason` on the [integrity test](https://github.com/pandres95/polkadot-sdk/blob/c4fabebf5fbd4878ac3ff1d60ba5670aa3351afa/substrate/frame/assets-freezer/src/lib.rs), but allowing the implementor of the pallet to use any other type as freeze identifier in case that's needed by the _runtime dev_.

Reply:
@muharem I was also curious and asked in Element but no one answered.

My best guess is it's to allow constraining the FreezeReasons allowed to be used in the pallet?

Reply:
Lets imagine `FreezeIdentifier` is some non enum type, `u32` for example. The overarching freeze reason in this case same - `u32`. The varian count some const `u32` number. 
So I would imagine this as 
```
trat Config {
    /// The overarching freeze reason.
    type RuntimeFreezeReason: Parameter + Member + MaxEncodedLen + Copy;
    /// The overarching freeze reason.
    type MaxFreezes: Get<u32>;
}
```
where any type that implements `VariantCount`, implements `Get<u32>` by default via generic impl.

Reply:
there is actually an adapter `VariantCountOf`

Reply:
Good catch! In `pallet-balances`, the struct `VariantCountOf` is used to constraint the limit of `Holds`, like this:

https://github.com/paritytech/polkadot-sdk/blob/0d71753e0eb1c0fdbeb09af5db7a2e1b5f150b13/substrate/frame/balances/src/lib.rs#L476-L485

This approach works in such a way there's no need to test for integrity, removes the need for a `MaxHolds`, and the ambiguity between `RuntimeFreezeReason` and `FreezeIdentifier`. I can apply the same strategy for `RuntimeFreezeReason` in `pallet-assets-freezer`.

**Open question:** considering a freeze reason can be of type `RuntimeFreezeReason` or any open-ended type that implements `VariantCount`, should it really be named `Runtime` in the type?

**Note:** I also suggest opening an issue to consider whether to standardize this approach to replace `FreezeIdentifier`/`MaxFreezes` in `pallet-balances` in favour of using `RuntimeFreezeReason` as the freezes Identifier there. Noticeably, for a potentially future implementation of `fungibles::hold` in `pallet-assets` either via a pallet extension, or via a direct implementation of those traits. would be good to apply the same approach.

Reply:
@muharem just [removed `FreezeIdentifier` in favour of `RuntimeFreezeReason`](https://github.com/paritytech/polkadot-sdk/pull/3951/commits/2b17fbbc354b1cd3a3541de9eba70152417ed4ca) as an unambiguous type for identifying freezes on the pallet. Left a comment above to explain the decision, and proposed some other questions and follow-up steps.